firestore/integration_test_internal/src/jni/arena_ref_test.cc

```diff
@@ -0,0 +1,278 @@
+  Local<String> string;
+  ArenaRef arena_ref(env(), string);
+  EXPECT_EQ(arena_ref.get(env()).get(), nullptr);
+}
+
+TEST_F(ArenaRefTestAndroid, ConstructsFromValid) {
+  Local<String> string = env().NewStringUtf("hello world");
+  ArenaRef arena_ref(env(), string);
+  EXPECT_TRUE(env().IsSameObject(arena_ref.get(env()), string));
+}
+
+TEST_F(ArenaRefTestAndroid, CopyConstructsFromNull) {
+  ArenaRef arena_ref1;
+  ArenaRef arena_ref2(arena_ref1);
+  EXPECT_EQ(arena_ref2.get(env()).get(), nullptr);
+}
+
+TEST_F(ArenaRefTestAndroid, CopyConstructsFromValid) {
+  Local<String> string = env().NewStringUtf("hello world");
+
+  ArenaRef arena_ref1(env(), string);
+  ArenaRef arena_ref2(arena_ref1);
+  EXPECT_TRUE(env().IsSameObject(arena_ref1.get(env()), string));
+  EXPECT_TRUE(env().IsSameObject(arena_ref2.get(env()), string));
+}
+
+TEST_F(ArenaRefTestAndroid, CopyAssignsFromNullToNull) {
+  ArenaRef arena_ref1, arena_ref2;
+  arena_ref2 = arena_ref1;
+  EXPECT_EQ(arena_ref1.get(env()).get(), nullptr);
+  EXPECT_EQ(arena_ref2.get(env()).get(), nullptr);
+}
+
+TEST_F(ArenaRefTestAndroid, CopyAssignsFromNullToValid) {
+  Local<String> string = env().NewStringUtf("hello world");
+
+  ArenaRef arena_ref1;
+  ArenaRef arena_ref2(env(), string);
+  arena_ref2 = arena_ref1;
+  EXPECT_EQ(arena_ref1.get(env()).get(), nullptr);
+  EXPECT_EQ(arena_ref2.get(env()).get(), nullptr);
+}
+
+TEST_F(ArenaRefTestAndroid, CopyAssignsFromValidToNull) {
+  Local<String> string = env().NewStringUtf("hello world");
+
+  ArenaRef arena_ref1;
+  ArenaRef arena_ref2(env(), string);
+  arena_ref1 = arena_ref2;
+  EXPECT_TRUE(env().IsSameObject(arena_ref1.get(env()), string));
+  EXPECT_TRUE(env().IsSameObject(arena_ref2.get(env()), string));
+}
+
+TEST_F(ArenaRefTestAndroid, CopyAssignsFromValidToValid) {
+  Local<String> string1 = env().NewStringUtf("hello world");
+  Local<String> string2 = env().NewStringUtf("hello earth");
+
+  ArenaRef arena_ref1(env(), string1);
+  ArenaRef arena_ref2(env(), string2);
+  arena_ref1 = arena_ref2;
+
+  EXPECT_TRUE(env().IsSameObject(arena_ref1.get(env()), string2));
+  EXPECT_TRUE(env().IsSameObject(arena_ref2.get(env()), string2));
+}
+
+TEST_F(ArenaRefTestAndroid, CopyAssignsFromNullObjectItself) {
+  ArenaRef arena_ref1;
+  arena_ref1 = arena_ref1;
+  EXPECT_EQ(arena_ref1.get(env()).get(), nullptr);
+}
+
+TEST_F(ArenaRefTestAndroid, CopyAssignsFromValidObjectItself) {
+  Local<String> string1 = env().NewStringUtf("hello world");
+
+  ArenaRef arena_ref1(env(), string1);
+  arena_ref1 = arena_ref1;
+  EXPECT_TRUE(env().IsSameObject(arena_ref1.get(env()), string1));
+}
+
+TEST_F(ArenaRefTestAndroid, MoveConstructsFromNull) {
+  ArenaRef arena_ref1;
+  ArenaRef arena_ref2(std::move(arena_ref1));
+  EXPECT_EQ(arena_ref2.get(env()).get(), nullptr);
+}
+
+TEST_F(ArenaRefTestAndroid, MoveConstructsFromValid) {
+  Local<String> string = env().NewStringUtf("hello world");
+
+  ArenaRef arena_ref2(env(), string);
```

Comment:
nit: Rename `arena_ref2` and `arena_ref3` back to `arena_ref1` and `arena_ref2`, respectively.